### ISSUE TITLE
[Bug] Installation of hydra_ray_launcher with Python 3.8 fails due to pickle5 error.

### ISSUE THREAD

**Issue by ncvv** [facebookresearch/hydra#2079]
# 🐛 Bug
## Description
I am having trouble installing `hydra-ray-launcher` on my Windows 10 machine using Python 3.8 due to an error while building pickle5 which is part of the plugin installation requirements (see [here](https://github.com/facebookresearch/hydra/blob/796fdf1c54eabdcd5121d3a9c8ab6af70ee41763/plugins/hydra_ray_launcher/setup.py#L33)).

## Checklist

- [x]  I checked on the latest version of Hydra.
- [x]  I created a minimal repro.

## To reproduce

```console
PS C:\Users\ncvv\src> python --version
Python 3.8.6
PS C:\Users\ncvv\src> python -m venv C:/Users/ncvv/venvs/hydra-launcher
PS C:\Users\ncvv\src> & C:/Users/ncvv/venvs/hydra-launcher/Scripts/Activate.ps1
(hydra-launcher) PS C:\Users\ncvv\src> pip install hydra-core --upgrade
Collecting hydra-core
  Using cached hydra_core-1.1.1-py3-none-any.whl (145 kB)
Collecting omegaconf==2.1.*
  Using cached omegaconf-2.1.1-py3-none-any.whl (74 kB)
Collecting importlib-resources; python_version < "3.9"
  Using cached importlib_resources-5.4.0-py3-none-any.whl (28 kB)
Processing c:\users\ncvv\appdata\local\pip\cache\wheels\c8\d0\ab\d43c02eaddc5b9004db86950802442ad9a26f279c619e28da0\antlr4_python3_runtime-4.8-py3-none-any.whl 
  Using cached PyYAML-6.0-cp38-cp38-win_amd64.whl (155 kB)
Collecting zipp>=3.1.0; python_version < "3.10"
  Using cached zipp-3.7.0-py3-none-any.whl (5.3 kB)
Installing collected packages: antlr4-python3-runtime, PyYAML, omegaconf, zipp, importlib-resources, hydra-core
Successfully installed PyYAML-6.0 antlr4-python3-runtime-4.8 hydra-core-1.1.1 importlib-resources-5.4.0 omegaconf-2.1.1 zipp-3.7.0
(hydra-launcher) PS C:\Users\ncvv\src> pip install hydra-ray-launcher --upgrade
Collecting hydra-ray-launcher
  Using cached hydra_ray_launcher-1.1.0-py3-none-any.whl (17 kB)
Requirement already satisfied: hydra-core>=1.1.0.dev7 in c:\users\ncvv\venvs\hydra-launcher\lib\site-packages (from hydra-ray-launcher) (1.1.1)
Collecting boto3==1.17.17
  Using cached boto3-1.17.17-py2.py3-none-any.whl (130 kB)
Collecting pickle5==0.0.11
  Using cached pickle5-0.0.11.tar.gz (132 kB)
  Preparing metadata (setup.py) ... done
Collecting ray==1.3.0
  Using cached ray-1.3.0-cp38-cp38-win_amd64.whl (15.5 MB)
Collecting cloudpickle==1.6.0
  Using cached cloudpickle-1.6.0-py3-none-any.whl (23 kB)
Collecting botocore<1.21.0,>=1.20.17
  Using cached botocore-1.20.112-py2.py3-none-any.whl (7.7 MB)
Collecting s3transfer<0.4.0,>=0.3.0
  Using cached s3transfer-0.3.7-py2.py3-none-any.whl (73 kB)
Collecting jmespath<1.0.0,>=0.7.1
  Using cached jmespath-0.10.0-py2.py3-none-any.whl (24 kB)
Collecting msgpack<2.0.0,>=1.0.0
  Using cached msgpack-1.0.3-cp38-cp38-win_amd64.whl (69 kB)
Collecting filelock
  Using cached filelock-3.6.0-py3-none-any.whl (10.0 kB)
Requirement already satisfied: pyyaml in c:\users\ncvv\venvs\hydra-launcher\lib\site-packages (from ray==1.3.0->hydra-ray-launcher) (6.0)
Collecting numpy>=1.16
  Downloading numpy-1.22.3-cp38-cp38-win_amd64.whl (14.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.7/14.7 MB 26.2 MB/s eta 0:00:00
Collecting prometheus-client>=0.7.1
  Downloading prometheus_client-0.13.1-py3-none-any.whl (57 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.1/57.1 KB ? eta 0:00:00
Collecting gpustat
  Using cached gpustat-0.6.0-py3-none-any.whl
Collecting aiohttp
  Using cached aiohttp-3.8.1-cp38-cp38-win_amd64.whl (555 kB)
Collecting py-spy>=0.2.0
  Using cached py_spy-0.3.11-py2.py3-none-win_amd64.whl (1.4 MB)
Collecting opencensus
  Using cached opencensus-0.8.0-py2.py3-none-any.whl (128 kB)
Collecting requests
  Downloading requests-2.27.1-py2.py3-none-any.whl (63 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.1/63.1 KB ? eta 0:00:00
Collecting redis>=3.5.0
  Using cached redis-4.1.4-py3-none-any.whl (175 kB)
Collecting aiohttp-cors
  Using cached aiohttp_cors-0.7.0-py3-none-any.whl (27 kB)
Collecting aioredis
  Using cached aioredis-2.0.1-py3-none-any.whl (71 kB)
Collecting grpcio>=1.28.1
  Downloading grpcio-1.44.0-cp38-cp38-win_amd64.whl (3.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 31.4 MB/s eta 0:00:00
Collecting jsonschema
  Downloading jsonschema-4.4.0-py3-none-any.whl (72 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.7/72.7 KB ? eta 0:00:00
Collecting protobuf>=3.15.3
  Downloading protobuf-3.19.4-cp38-cp38-win_amd64.whl (895 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 895.9/895.9 KB 28.6 MB/s eta 0:00:00
Collecting colorama
  Using cached colorama-0.4.4-py2.py3-none-any.whl (16 kB)
Collecting click>=7.0
  Using cached click-8.0.4-py3-none-any.whl (97 kB)
Requirement already satisfied: antlr4-python3-runtime==4.8 in c:\users\ncvv\venvs\hydra-launcher\lib\site-packages (from hydra-core>=1.1.0.dev7->hydra-ray-launcher) (4.8)
Requirement already satisfied: importlib-resources in c:\users\ncvv\venvs\hydra-launcher\lib\site-packages (from hydra-core>=1.1.0.dev7->hydra-ray-launcher) (5.4.0)
Requirement already satisfied: omegaconf==2.1.* in c:\users\ncvv\venvs\hydra-launcher\lib\site-packages (from hydra-core>=1.1.0.dev7->hydra-ray-launcher) (2.1.1)
Collecting urllib3<1.27,>=1.25.4
  Downloading urllib3-1.26.8-py2.py3-none-any.whl (138 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 138.7/138.7 KB 8.0 MB/s eta 0:00:00
Collecting python-dateutil<3.0.0,>=2.1
  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
Collecting six>=1.5.2
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting packaging>=20.4
  Using cached packaging-21.3-py3-none-any.whl (40 kB)
Collecting deprecated>=1.2.3
  Using cached Deprecated-1.2.13-py2.py3-none-any.whl (9.6 kB)
Collecting attrs>=17.3.0
  Downloading attrs-21.4.0-py2.py3-none-any.whl (60 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 KB ? eta 0:00:00
Collecting frozenlist>=1.1.1
  Using cached frozenlist-1.3.0-cp38-cp38-win_amd64.whl (33 kB)
Collecting aiosignal>=1.1.2
  Using cached aiosignal-1.2.0-py3-none-any.whl (8.2 kB)
Collecting yarl<2.0,>=1.0
  Using cached yarl-1.7.2-cp38-cp38-win_amd64.whl (122 kB)
Collecting async-timeout<5.0,>=4.0.0a3
  Using cached async_timeout-4.0.2-py3-none-any.whl (5.8 kB)
Collecting charset-normalizer<3.0,>=2.0
  Using cached charset_normalizer-2.0.12-py3-none-any.whl (39 kB)
Collecting multidict<7.0,>=4.5
  Using cached multidict-6.0.2-cp38-cp38-win_amd64.whl (28 kB)
Collecting typing-extensions
  Downloading typing_extensions-4.1.1-py3-none-any.whl (26 kB)
Collecting nvidia-ml-py3>=7.352.0
  Using cached nvidia_ml_py3-7.352.0-py3-none-any.whl
Collecting blessings>=1.6
  Using cached blessings-1.7-py3-none-any.whl (18 kB)
Collecting psutil
  Downloading psutil-5.9.0-cp38-cp38-win_amd64.whl (246 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 246.5/246.5 KB 15.7 MB/s eta 0:00:00
Requirement already satisfied: zipp>=3.1.0 in c:\users\ncvv\venvs\hydra-launcher\lib\site-packages (from importlib-resources->hydra-core>=1.1.0.dev7->hydra-ray-launcher) (3.7.0)
Collecting pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0
  Downloading pyrsistent-0.18.1-cp38-cp38-win_amd64.whl (61 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.7/61.7 KB ? eta 0:00:00
Collecting google-api-core<3.0.0,>=1.0.0
  Downloading google_api_core-2.7.0-py3-none-any.whl (114 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.7/114.7 KB ? eta 0:00:00
Collecting opencensus-context==0.1.2
  Using cached opencensus_context-0.1.2-py2.py3-none-any.whl (4.4 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.3-py3-none-any.whl (61 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2021.10.8-py2.py3-none-any.whl (149 kB)
Collecting wrapt<2,>=1.10
  Using cached wrapt-1.13.3-cp38-cp38-win_amd64.whl (34 kB)
Collecting googleapis-common-protos<2.0dev,>=1.52.0
  Downloading googleapis_common_protos-1.55.0-py2.py3-none-any.whl (212 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 212.3/212.3 KB ? eta 0:00:00
Collecting google-auth<3.0dev,>=1.25.0
  Downloading google_auth-2.6.0-py2.py3-none-any.whl (156 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 156.3/156.3 KB ? eta 0:00:00
Collecting pyparsing!=3.0.5,>=2.0.2
  Downloading pyparsing-3.0.7-py3-none-any.whl (98 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.0/98.0 KB ? eta 0:00:00
Collecting pyasn1-modules>=0.2.1
  Using cached pyasn1_modules-0.2.8-py2.py3-none-any.whl (155 kB)
Collecting cachetools<6.0,>=2.0.0
  Downloading cachetools-5.0.0-py3-none-any.whl (9.1 kB)
Collecting rsa<5,>=3.1.4
  Using cached rsa-4.8-py3-none-any.whl (39 kB)
Collecting pyasn1<0.5.0,>=0.4.6
  Using cached pyasn1-0.4.8-py2.py3-none-any.whl (77 kB)
Using legacy 'setup.py install' for pickle5, since package 'wheel' is not installed.
Installing collected packages: pyasn1, py-spy, pickle5, opencensus-context, nvidia-ml-py3, msgpack, certifi, wrapt, urllib3, typing-extensions, six, rsa, pyrsistent, pyparsing, pyasn1-modules, psutil, protobuf, prometheus-client, numpy, multidict, jmespath, idna, frozenlist, filelock, colorama, cloudpickle, charset-normalizer, cachetools, attrs, async-timeout, yarl, requests, python-dateutil, packaging, jsonschema, grpcio, googleapis-common-protos, google-auth, deprecated, click, blessings, aiosignal, aioredis, redis, gpustat, google-api-core, botocore, aiohttp, s3transfer, opencensus, aiohttp-cors, ray, boto3, hydra-ray-launcher
  Running setup.py install for pickle5 ... error
  error: subprocess-exited-with-error

  × Running setup.py install for pickle5 did not run successfully.
  │ exit code: 1
  ╰─> [36 lines of output]
      running install
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-3.8
      creating build\lib.win-amd64-3.8\pickle5
      copying pickle5\pickle.py -> build\lib.win-amd64-3.8\pickle5
      copying pickle5\pickletools.py -> build\lib.win-amd64-3.8\pickle5
      copying pickle5\__init__.py -> build\lib.win-amd64-3.8\pickle5
      creating build\lib.win-amd64-3.8\pickle5\test
      copying pickle5\test\pickletester.py -> build\lib.win-amd64-3.8\pickle5\test
      copying pickle5\test\test_pickle.py -> build\lib.win-amd64-3.8\pickle5\test
      copying pickle5\test\test_picklebuffer.py -> build\lib.win-amd64-3.8\pickle5\test
      copying pickle5\test\__init__.py -> build\lib.win-amd64-3.8\pickle5\test
      running build_ext
      building 'pickle5._pickle' extension
      creating build\temp.win-amd64-3.8
      creating build\temp.win-amd64-3.8\Release
      creating build\temp.win-amd64-3.8\Release\pickle5
      C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Ic:\users\ncvv\venvs\hydra-launcher\include -IC:\Users\ncvv\Python38\include -IC:\Users\ncvv\Python38\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.16.27023\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.16.27023\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt" /Tcpickle5/_pickle.c /Fobuild\temp.win-amd64-3.8\Release\pickle5/_pickle.obj
      _pickle.c
      C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Ic:\users\ncvv\venvs\hydra-launcher\include -IC:\Users\ncvv\Python38\include -IC:\Users\ncvv\Python38\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.16.27023\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.16.27023\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt" /Tcpickle5/picklebufobject.c /Fobuild\temp.win-amd64-3.8\Release\pickle5/picklebufobject.obj
      picklebufobject.c
      pickle5/picklebufobject.c(20): warning C4273: 'PyPickleBuffer_FromObject': inconsistent dll linkage
      c:\users\ncvv\python38\include\picklebufobject.h(18): note: see previous definition of 'PyPickleBuffer_FromObject'
      pickle5/picklebufobject.c(39): warning C4273: 'PyPickleBuffer_GetBuffer': inconsistent dll linkage
      c:\users\ncvv\python38\include\picklebufobject.h(22): note: see previous definition of 'PyPickleBuffer_GetBuffer'
      pickle5/picklebufobject.c(58): warning C4273: 'PyPickleBuffer_Release': inconsistent dll linkage
      c:\users\ncvv\python38\include\picklebufobject.h(24): note: see previous definition of 'PyPickleBuffer_Release'
      pickle5/picklebufobject.c(208): warning C4273: 'PyPickleBuffer_Type': inconsistent dll linkage
      c:\users\ncvv\python38\include\picklebufobject.h(13): note: see previous definition of 'PyPickleBuffer_Type'
      C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:c:\users\ncvv\venvs\hydra-launcher\libs /LIBPATH:C:\Users\ncvv\Python38\libs /LIBPATH:C:\Users\ncvv\Python38 /LIBPATH:c:\users\ncvv\venvs\hydra-launcher\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.16.27023\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.16.27023\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.19041.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.19041.0\um\x64" /EXPORT:PyInit__pickle build\temp.win-amd64-3.8\Release\pickle5/_pickle.obj build\temp.win-amd64-3.8\Release\pickle5/picklebufobject.obj /OUT:build\lib.win-amd64-3.8\pickle5\_pickle.cp38-win_amd64.pyd /IMPLIB:build\temp.win-amd64-3.8\Release\pickle5\_pickle.cp38-win_amd64.lib
      python38.lib(python38.dll) : error LNK2005: PyPickleBuffer_GetBuffer already defined in picklebufobject.obj
         Creating library build\temp.win-amd64-3.8\Release\pickle5\_pickle.cp38-win_amd64.lib and object build\temp.win-amd64-3.8\Release\pickle5\_pickle.cp38-win_amd64.exp
      build\lib.win-amd64-3.8\pickle5\_pickle.cp38-win_amd64.pyd : fatal error LNK1169: one or more multiply defined symbols found
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Professional\\VC\\Tools\\MSVC\\14.16.27023\\bin\\HostX86\\x64\\link.exe' failed with exit status 1169
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> pickle5

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
```

## Expected Behavior
AFAIU, pickle5 is supposed to backport the pickle protocol version 5 to Python 3.5, 3.6 and 3.7. Since this new protocol version was merged into Python 3.8 anyways and hydra-ray-launcher supports installation with Python 3.8, I would expect that pickle5 is not a necessary requirement for the plugin but rather an extra requirement to ensure backwards compatibility with older Python versions such as 3.7.

## System information
Hydra Version : 1.1.1
Python version : 3.8.6
OS : Windows 10 Pro, Version 21H1, Build 19043.1526

**Comment by Jasha10** [comment]
Thanks for the report @ncvv.

Perhaps updating the pin `pickle5==0.0.11` to `pickle5==0.0.12` would solve this
problem. It seems that, since v0.0.11, the `pickle5` repository has merged a [pull request](https://github.com/pitrou/pickle5-backport/pull/26) pertaining specifically to later python versions such as v3.8.

You can try this out locally as follows:
```
git clone https://github.com/facebookresearch/hydra
cd hydra/plugins/hydra_ray_launcher
# modify hydra/plugins/hydra_ray_launcher/setup.py
pip install .
```
If that works on your system, a PR would be welcome!

**Comment by Jasha10** [comment]
If replacing `"pickle5==0.0.11"` with `"pickle5==0.0.12"` does not work, please also try the following:
```'pickle5==0.0.11; python_version < "3.8.3"'```

**Comment by jieru-hu** [comment]
thanks @Jasha10  for taking a look.

Thanks for the issue @ncvv ! We officially do not support ray launcher on windows yet (at the time of developing ray launcher, ray does not support Windows. It seems ray has since added support for windows.)
This also means we **do not have any testing coverage** for ray launcher on windows. 
I've created https://github.com/facebookresearch/hydra/issues/2086 to track this.

**Comment by ncvv** [comment]
(After reading jieru-hu's answer, I am just stating what I had already done for completeness)

@Jasha10 Updating the pin does not solve the issue because in the PR you mentioned, pickle5 specifically introduces `python_requires='>=3.5, <3.8'` to avoid installation with Python 3.8. I've also tried adding `'pickle5==0.0.11; python_version < "3.8.3"'` to the hydra-ray-launcher setup.py which makes the installation work, but then the tests fail:
```
ERROR tests/test_ray_aws_launcher.py - pkg_resources.DistributionNotFound: The 'pickle5' distribution was not found and is required by the application
``` 

@jieru-hu Thanks for your reply, I did not know that. Will try the launcher using linux. Thanks for adding the feature request for windows support! Closing this for now.
